cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramNotificationSubscriberService.java

```diff
@@ -223,35 +230,65 @@ private void processNotification(AppMetadataStore appMetadataStore, Notification
                      programRunId, ProgramOptionConstants.LOGICAL_START_TIME, notification);
             return;
           }
-          appMetadataStore.recordProgramRunning(programId, runId, logicalStartTimeSecs, twillRunId, messageIdBytes);
-          return;
+          recorded =
+            appMetadataStore.recordProgramRunning(programId, runId, logicalStartTimeSecs, twillRunId, messageIdBytes);
+          break;
         case SUSPENDED:
-          appMetadataStore.recordProgramSuspend(programId, runId, messageIdBytes);
-          return;
+          recorded = appMetadataStore.recordProgramSuspend(programId, runId, messageIdBytes);
+          break;
         case RESUMING:
-          appMetadataStore.recordProgramResumed(programId, runId, messageIdBytes);
-          return;
+          recorded = appMetadataStore.recordProgramResumed(programId, runId, messageIdBytes);
+          break;
         case COMPLETED:
         case KILLED:
           if (endTimeSecs == -1) {
             LOG.warn("Ignore program killed notification for program {} without end time specified, {}",
                      programRunId, notification);
             return;
           }
-          appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, null, messageIdBytes);
-          return;
+          recorded =
+            appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, null, messageIdBytes);
+          break;
         case FAILED:
           if (endTimeSecs == -1) {
             LOG.warn("Ignore program failed notification for program {} without end time specified, {}",
                      programRunId, notification);
             return;
           }
           BasicThrowable cause = decodeBasicThrowable(properties.get(ProgramOptionConstants.PROGRAM_ERROR));
-          appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, cause, messageIdBytes);
-          return;
+          recorded =
+            appMetadataStore.recordProgramStop(programId, runId, endTimeSecs, programRunStatus, cause, messageIdBytes);
+          break;
         default:
           // This should not happen
           LOG.error("Unsupported program status {} for program {}, {}", programRunStatus, programRunId, notification);
+          return;
+      }
+      if (recorded) {
+        // Notification with status RESUMING is actually recorded as RUNNING
```

Comment:
Why having the logic in here? I think the `appMetaStore.record` method should return the state change.

Reply:
if it fails to record, just return `null`?

Reply:
What do mean by failed to record? You mean no update, right? If you define the method to return the actual state being written, then yes, you can use `null` to represent no update. If you need more information that just the state, you can also have an object carrying those info.